### PR TITLE
Separate timeout for handler sync from informer sync and do not resync services during node tracker startup

### PR DESCRIPTION
**Pull request by ricky-rav**
[edited]
- separate the handler sync from the informer sync 
-  remove the full service resync during the node tracker startup since it's redundant: all service keys are added at each node tracker add, but no service controller workers are consuming those keys yet; the services keys are added already in the service controller startup. This is discussed in: https://issues.redhat.com/browse/OCPBUGS-20336?focusedId=23916634&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-23916634 and previous comments

### CODE REVIEW

File: go-controller/pkg/types/const.go
Comment (tssurya) [inline go-controller/pkg/types/const.go:204]: how long did it take on the scale cluster? Could you provide some more data around the size of the cluster? (nodes, services)? What if 2mins is also not enough or do we think 2mins has to be enough.?

File: go-controller/pkg/util/sync.go
Comment (tssurya) [inline go-controller/pkg/util/sync.go:31]: hmm do you think we should rename this util from 
`func WaitForNamedCacheSyncWithTimeout(controllerName string, stopCh <-chan struct{}, cacheSyncs ...cache.InformerSynced) bool {`
to 
`func WaitForInformerCacheSyncWithTimeout(controllerName string, stopCh <-chan struct{}, cacheSyncs ...cache.InformerSynced) bool {`

so that its explicit ? NamedCacheSync seems too generic?

**Review by tssurya — CHANGES_REQUESTED** [review]

File: go-controller/pkg/types/const.go
Comment (ricky-rav) [reply go-controller/pkg/types/const.go:204]: You're right, I'll add that to the description.

File: go-controller/pkg/util/sync.go
Comment (ricky-rav) [reply go-controller/pkg/util/sync.go:31]: True

**Comment by npinaeva** [comment]
@ricky-rav can you also update handlers to informers here please (I am sure in the first case I intended to sync informer, not sure which sync we needed in the second case)
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/controller/controller.go#L83
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/controller/services/services_controller.go#L198

**Comment by tssurya** [comment]
> In short, we saw that with a 250-node cluster it takes node tracker an average of 0.1 seconds to process one node, so in the upgrade runs that we looked at, the time out of 2 minutes occurred when ~200 nodes had been processed.

was it 2mins or 20sec? based on your presentation seems 20sec?

**Comment by ricky-rav** [comment]
I have two commits now:
- I separate the handler sync from the informer sync and I keep the same timeout of 20s in both cases (here I was increasing the handler sync timeout to 2 minutes before)
- I remove the full service resync during the node tracker startup, as discussed in last couple of comments of this bug: https://issues.redhat.com/browse/OCPBUGS-20336?focusedId=23916634&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-23916634

I have less evidence now to increase the handler timeout sync to 2 minutes, since I've optimized the node tracker startup code in the second commit and that part should now take ~0.03 in a 250-node cluster. Do we want to still do it? The handlers for which we call `WaitForHandlerSyncWithTimeout` are all level-driven controllers, so `WaitForHandlerSyncWithTimeout` will return when all object keys have been added to the queue, without waiting for the workers to process the keys. That shouldn't need a higher timeout than 20s. 

I think it still makes sense to have `WaitForHandlerSyncWithTimeout` and `WaitForInformerSyncWithTimeout`, because they wait for two different things, but the reasons for a higher timeout are thinner now. @npinaeva @tssurya @trozet what do you all think?

Thanks!

**Comment by ricky-rav** [comment]
/hold
There's a nice race condition in the variable I added...

**Comment by ricky-rav** [comment]
Should be good to go now

**Comment by ricky-rav** [comment]
> May be we should make `WaitForHandlerSyncWithTimeout` always accept the timeout explicitly from the caller, considering there is not such thing as "generic handler timeout" since it depends on the handler implementation? Then you could tweak it per-handler as needed, and also make sure you do some scale testing before setting it.

Yeah, that's not a bad idea to make it explicit. I've set all three timeouts to 20 seconds, which seems to have been a good upper bound so far.

File: go-controller/pkg/controller/controller.go
Comment (npinaeva) [inline go-controller/pkg/controller/controller.go:84]: oh I meant to change this to informer sync, so use `c.config.Informer.HasSynced` instead of `c.eventHandler.HasSynced`

Comment (ricky-rav) [reply go-controller/pkg/controller/controller.go:84]: Oh I see, I read your other message too quickly.

-  in services_controller.go (https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/controller/services/services_controller.go#L198) I definitely need `WaitForHandlerSyncWithTimeout`, because I want all service keys to be added to the queue at that step, since I removed that part from the node tracker startup.

- here in controller.go the existing code passes the handler `HasSync`, so I simply figured I'd keep the existing logic and use `WaitForHandlerSyncWithTimeout`. If that's not necessary, I can definitely change that.

File: go-controller/pkg/controller/controller.go
Comment (npinaeva) [inline go-controller/pkg/controller/controller.go:84]: lol, nice catch!

**Comment by npinaeva** [comment]
lgtm for the sync wait logic, thanks @ricky-rav!

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (tssurya) [inline go-controller/pkg/ovn/controller/services/services_controller.go:175]: qq: we were not syncing services here anyways right? you have just added this comment to explicitly call that out, that's all right?

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (tssurya) [inline go-controller/pkg/ovn/controller/services/services_controller.go:532]: shouldn't this be .... return the error and fail hard/retry? or do we want to keep it as is..

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (tssurya) [inline go-controller/pkg/ovn/controller/services/services_controller.go:525]: so.. here the ordering is:
initialize: svccontroller.NewController->NewNodeTRacker->RequestFullSync->we check if initialNTSync is done which won't be done since that part of code is not touched yet right?

Then we have the:
Run code => after the initialize is done and we have the DNC, we do DNC.Run-> startSvcController->Run-> do the initial node tracker sync

I am unsure how and when we can hit RequestFullSync.NTInitialSync=true condition? It seems like the initial sync is always going to happen afterwards right? Perhaps I am missing something. I will huddle with you.

File: go-controller/pkg/util/sync.go
Comment (tssurya) [reply go-controller/pkg/util/sync.go:34]: resolved thanks!

File: go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go
Comment (ricky-rav) [reply go-controller/pkg/clustermanager/egressservice/egressservice_cluster.go:176]: ah ops, there were a few of them :-D

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/services_controller.go:175]: we were doing it before, but now that I removed it, the comment alone doesn't tell you the whole story. Let me remove it from here and add it to `RequestFullSync`, where it makes more sense.

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/services_controller.go:532]: You're right, I simply kept the existing logic. If I look at how node tracker sync works, it has no retry mechanism at all o_O
It might be something to explore for our tech debt, maybe?

Comment (tssurya) [reply go-controller/pkg/ovn/controller/services/services_controller.go:532]: yeah sounds good, deff don't want you to go out of scope, perhaps open a JIRA tracker for this please! thanks

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/services_controller.go:526]: I've moved the line where I set the flag to true further below and renamed it more simply `startupDone`, since it actually refers to whether the initial sync for the three handlers we have (node tracker, service, endpointslice) has been completed. I didn't move `c.startupDone=true` _after_ we start the workers because I'm a bit paranoid about the fact that we might miss node tracker events in the teeny tiny time frame of the for loop itself:
```
	klog.Info("Starting workers")
	for i := 0; i < workers; i++ {
		go wait.Until(c.worker, c.workerLoopPeriod, stopCh)
	}
```

**Comment by tssurya** [comment]
the two flakes in CI are known issues

**Review by tssurya — CHANGES_REQUESTED** [review]
both commits lgtm except for the nits I outlined

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (tssurya) [reply go-controller/pkg/ovn/controller/services/services_controller.go:526]: `I didn't move c.startupDone=true after we start the workers because I'm a bit paranoid about the fact that we might miss node tracker events in the teeny tiny time frame of the for loop itself`

ack that makes sense, but then there is also the reverse scene where for some nodes we get the add/update whatever between this tiny window where the done = true and workers are NOT Ready yet due to the loop not having finished but I think that should be negligible.

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/services_controller.go:222]: yesss

**Review by tssurya — APPROVED** [review]
latest changes LGTM.
d/s scale test has been done https://github.com/openshift/ovn-kubernetes/pull/2014 no disruptions or anything significant found yet.
once the CI runs here on the latest will merge this.
thanks @ricky-rav for the detailed commit message which made it easier for me to understand and your awesome troubleshooting here

**Comment by tssurya** [comment]
new flake in external-gateway lane which is not related to this PR: https://github.com/ovn-org/ovn-kubernetes/issues/4139
known flake in kv libe migration lane, rest are green: https://github.com/ovn-org/ovn-kubernetes/issues/3986

File: go-controller/pkg/ovn/controller/services/services_controller.go
Comment (ricky-rav) [reply go-controller/pkg/ovn/controller/services/services_controller.go:534]: Here it is: https://issues.redhat.com/browse/SDN-4483

**Pull request merged**
